File: skypydb-js/src/database/mixins/vector/delete_items.ts
import { CollectionNotFoundError } from "../../../errors";
import { InputValidator } from "../../../security/validation";
import { QueryItemsMixin } from "./query_items";

export abstract class DeleteItemsMixin extends QueryItemsMixin {
  delete(
    collection_name: string,
    ids?: string[],
    where?: Record<string, unknown>,
    where_document?: Record<string, string>
  ): number {
    const validated = InputValidator.validate_table_name(collection_name);
    if (!this.collection_exists(validated)) {
      throw new CollectionNotFoundError(`Collection '${validated}' not found`);
    }

    if (ids) {
      const placeholders = ids.map(() => "?").join(", ");
      const result = this.conn
        .prepare(`DELETE FROM [vec_${validated}] WHERE id IN (${placeholders})`)
        .run(...ids);
      return Number(result.changes ?? 0);
    }

    const ids_to_delete: string[] = [];
    for (const item of this._get_all_items(validated)) {
      if (this._matches_filters(item, where, where_document)) {
        ids_to_delete.push(item.id);
      }
    }

    if (ids_to_delete.length === 0) {
      return 0;
    }

    const placeholders = ids_to_delete.map(() => "?").join(", ");
    const result = this.conn
      .prepare(`DELETE FROM [vec_${validated}] WHERE id IN (${placeholders})`)
      .run(...ids_to_delete);
    return Number(result.changes ?? 0);
  }
}

File: skypydb/database/operations/vector/delete_items.py
"""Delete operations for vector collection records."""

from typing import Any, Dict, List, Optional

from skypydb.security.validation import InputValidator


class DeleteItemsMixin:
    """Provide record deletion for vector collections."""

    def delete(
        self,
        collection_name: str,
        ids: Optional[List[str]] = None,
        where: Optional[Dict[str, Any]] = None,
        where_document: Optional[Dict[str, str]] = None,
    ) -> int:
        """Delete records by explicit IDs and/or filters."""

        collection_name = InputValidator.validate_table_name(collection_name)
        if not self.collection_exists(collection_name):
            raise ValueError(f"Collection '{collection_name}' not found")

        cursor = self.conn.cursor()

        if ids is not None:
            placeholders = ", ".join(["?" for _ in ids])
            cursor.execute(
                f"DELETE FROM [vec_{collection_name}] WHERE id IN ({placeholders})",
                list(ids),
            )
        else:
            ids_to_delete = []
            for item in self._get_all_items(collection_name):
                if self._matches_filters(item, where, where_document):
                    ids_to_delete.append(item["id"])

            if ids_to_delete:
                placeholders = ", ".join(["?" for _ in ids_to_delete])
                cursor.execute(
                    f"DELETE FROM [vec_{collection_name}] WHERE id IN ({placeholders})",
                    ids_to_delete,
                )

        deleted_count = cursor.rowcount
        self.conn.commit()
        return deleted_count
